x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -2,62 +2,84 @@ package keeper
+
 	common "github.com/dymensionxyz/dymension/v3/x/common/types"
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
-	"github.com/osmosis-labs/osmosis/v15/osmoutils"
 )
 
 // FinalizeRollappStates is called every block to finalize states when their dispute period over.
-func (k Keeper) FinalizeRollappStates(ctx sdk.Context) error {
+func (k Keeper) FinalizeRollappStates(ctx sdk.Context) {
 	if uint64(ctx.BlockHeight()) < k.DisputePeriodInBlocks(ctx) {
-		return nil
+		return
 	}
 	// check to see if there are pending  states to be finalized
 	finalizationHeight := uint64(ctx.BlockHeight() - int64(k.DisputePeriodInBlocks(ctx)))
 	pendingFinalizationQueue := k.GetAllFinalizationQueueUntilHeightInclusive(ctx, finalizationHeight)
 
-	return osmoutils.ApplyFuncIfNoError(ctx,
-		// we trap at this granularity because we want to avoid iterating inside the
-		// wrapped function (above), and we want the function to be atomic, so
-		// we cannot trap inside finalizePending
-		func(ctx sdk.Context) error {
-			return k.finalizePending(ctx, pendingFinalizationQueue)
-		})
+	k.finalizePending(ctx, pendingFinalizationQueue)
 }
 
-func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) error {
+func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) {
 	// Iterate over all the pending finalization queue
 	for _, blockHeightToFinalizationQueue := range pendingFinalizationQueue {
 		// finalize pending states
+		rollappsToRemoveFromQueue := make([]string, 0, len(blockHeightToFinalizationQueue.FinalizationQueue))
 		for _, stateInfoIndex := range blockHeightToFinalizationQueue.FinalizationQueue {
-			stateInfo := k.MustGetStateInfo(ctx, stateInfoIndex.RollappId, stateInfoIndex.Index)
-			if stateInfo.Status != common.Status_PENDING {
-				panic(fmt.Sprintf("invariant broken: stateInfo is not in pending state: rollapp: %s: status: %s", stateInfoIndex.RollappId, stateInfo.Status))
-			}
-			stateInfo.Finalize()
-			// update the status of the stateInfo
-			k.SetStateInfo(ctx, stateInfo)
-			// update the LatestStateInfoIndex of the rollapp
-			k.SetLatestFinalizedStateIndex(ctx, stateInfoIndex)
-			// call the after-update-state hook
-			keeperHooks := k.GetHooks()
-			err := keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
+			// if this fails, no state change will happen
+			err := osmoutils.ApplyFuncIfNoError(ctx,
+				func(ctx sdk.Context) error {
+					return k.finalizePendingState(ctx, stateInfoIndex)
+				})
 			if err != nil {
-				return fmt.Errorf("after state finalized: %w", err)
+				// if a rollapp's state fails to finalize, move on (state index will stay in the queue)
+				// TODO: think about (non)recoverable errors and how to handle them accordingly
+				k.Logger(ctx).Error("failed to finalize state", "rollapp", stateInfoIndex.RollappId, "index", stateInfoIndex.Index, "error", err.Error())
+				continue
```

Comment:
can it cause invariant break?
if some states of a rollapp are finalized but some don't